src/translator/quality_estimator.h

```diff
@@ -0,0 +1,152 @@
+class QualityEstimator {
+ public:
+  struct Header {
+    uint64_t magic;             // BINARY_QE_MODEL_MAGIC
+    uint64_t lrParametersDims;  // Length of lr parameters stds, means and coefficients .
+  };
+
+  /// WordsQualityEstimate contains the quality data of a given translated sentence.
+  /// It includes the confidence (proxied by a probability) of each decoded word
+  /// (higher probabilities imply better-translated words), the ByteRanges of each term,
+  /// and the probability of the whole sentence, represented as the mean word scores.
+  struct WordsQualityEstimate {
+    std::vector<float> wordQualityScores;
+    std::vector<ByteRange> wordByteRanges;
+    float sentenceScore = 0.0;
+  };
+
+  /// Construct a QualityEstimator
+  /// @param [in] qualityEstimatorMemory: AlignedMemory built from quality estimator binary file
+  explicit QualityEstimator(const AlignedMemory &qualityEstimatorMemory);
+
+  /// construct the struct WordsQualityEstimate
+  /// @param [in] logProbs: the log probabilities given by an translation model
+  /// @param [in] target: AnnotatedText target value
+  /// @param [in] sentenceIdx: the id of a candidate sentence
+  WordsQualityEstimate computeQualityScores(const std::vector<float> &logProbs, const AnnotatedText &target,
+                                            const size_t sentenceIdx) const;
+
+  AlignedMemory toAlignedMemory() const;
+
+ private:
+  struct Matrix {
+    Matrix(const size_t rowsParam, const size_t widthParam);
+
+    const float &at(const size_t row, const size_t col) const;
+    float &at(const size_t row, const size_t col);
+
+    const size_t rows;
+    const size_t cols;
+    AlignedVector<float> data;
+  };
+
+  /// A strcut to simplify the usage of intgemm
+  struct IntgemmMatrix : Matrix {
+    IntgemmMatrix(const intgemm::Index rowsParam, const intgemm::Index widthParam, const intgemm::Index rowsMultiplier,
+                  const intgemm::Index widthMultiplier);
+
+     Matrix operator*( const IntgemmMatrix& matrixb ) const;
+  };
+
+  struct Scale {
+    std::vector<float> stds;
+    std::vector<float> means;
+  };
+
+  /// WordFeatures represents N-1 features that are used by a given model.
+  ///
+  /// It's values are filled through mapBPEToWords
+  struct WordFeatures {
+    int numSubWords = 0;
+    float meanScore = 0.0;
+    float minScore = 0.0;
+  };
+
+  /// The current Quality Estimator model is a Logistic Model implemented through
+  /// a linear regressor + sigmoid function. Simply speaking, an LR model depends
+  /// on features to be scaled, so it contains four elements of data: a vector of coefficients
+  /// and an intercept (which represents the linear model) and a vector of means and stds
+  /// (which are necessary for feature scaling).
+  ///
+  /// These variables are firstly initialized by parsing a file (which comes from memory),
+  /// and then they are used to build a model representation
+  class LogisticRegressor {
```

Comment:
This class should exist outside QE.

Reply:
Considering that the LR is used only by QE ( at least afaik ) and QE will awalys have a model associated with it. Why should we make it public?

Reply:
Commit 88c41e7